add pieces sdk workshop (#9)
pieces-app/workshops: build-with-pieces-sdk/slides.md

---
layout: intro
introImage: '/img/pieces-circle.png'
transition: fade
---

# Build AI-powered Apps Using Pieces SDK

Learn how to use the Pieces OS Python SDK

<!--
Welcome to this workshop on learning how to use the pieces_os_client Python SDK.

The pieces_os_client Python SDK is a powerful tool for integrating Pieces OS functionality into your Python applications. This workshop will help you understand and utilize its key features.
-->

---
transition: fade
layout: center
---

> # What is Pieces?

Reviewer: Maybe add a QR code/link here so folks can start downloading it whilst the session is running?

<img class='bottom-image' width='400' src='/build-with-pieces-sdk/img/Robot_CTA_Coffee_DM.webp'/>

<!--

Pieces for Developers is an AI-powered tool designed to help developers, from students and hobbyists to professional developers, and even other users streamline their workflow and boost productivity. 

Unlike typical AI coding tools, Pieces is designed to understand the entire context of your workflow. This means you can share code snippets, code files and project folders, and even have it interact with all the apps that are running on your computer.

> Let's look at using Pieces in a variety of scenarios:

Reviewer: This is lifted from the student workshop, can you put more relevant speaker notes here please.

- How it can assist you in the classroom and lectures
- How you can use it for self guided learning
- Using it for hackathons
- Preparing for exams and interviews

-->

---
transition: fade
---

> # What is the Pieces OS SDK and how does it work?

Reviewer: Add a slide before explaining Pieces OS and what it is/how it fits in with Pieces. Would give more context to this slide.

The Pieces OS Client SDK is a powerful package that allows developers to build applications on top of Pieces OS. The SDK facilitates communication with a locally hosted Pieces OS server to enable features like copilot chats, asset saving, and more. 

We currently have Pieces OS Client SDKs for Python, JavaScript, Dart and Kotlin. 

<!--

Pieces OS Client SDK is a powerful package that allows developers to build applications on top of Pieces OS. The SDK facilitates communication with a locally hosted Pieces OS server to enable features like copilot chats, asset saving, and more.

Requirements: 
Pieces OS must be running as a backend service

-->

---
transition: fade
---

# What type of applications can be built with Pieces OS SDK?

> There are endless possibilities of what you can build with Pieces OS SDK. You can leverage the LLMs provided by Pieces OS to create conversational AI assistants, or use Pieces OS to integrate with your existing apps. Here are some potential examples for apps you build using Pieces OS SDKs:

Reviewer: Could also be fun to add that some of our open source plugins are using this SDK

<v-clicks>

- Conversational AI assistant powered by Pieces OS's local machine learning models
- Application that assists developers during code reviews by integrating with the Pieces OS. 
- Tool that automatically generates documentation based on the codebase and comments. The application can utilize Pieces OS to analyze code and produce structured documentation.

</v-clicks>

<!--

- Use the SDK to create a conversational AI assistant powered by Pieces OS's local machine learning models
- Create an application that assists developers during code reviews by integrating with the Pieces OS. 
- Create a tool that automatically generates documentation based on the codebase and comments. The application can utilize Pieces OS to analyze code and produce structured documentation.

-->

---
transition: fade
---

# Installation of Pieces OS Client SDK for Python

Install the SDK using pip:

```

> pip install pieces-os-client

Reviewer: `pieces_os_client` surely?

```

<!--
To get started, you'll need to install the SDK using pip. This command will download and install the necessary packages.
-->

---
transition: fade
---

# Core Elements

The `pieces_os_client` SDK consists of:

- **API Client**: Essential for all requests, storing crucial data such as the host.
- **Models**: Define the structure of the response body sent to the Pieces OS API.
- **APIs**: Modules offering diverse functionalities (e.g., AssetsAPI, QGPTApi).

<!--
The SDK is composed of three main elements:
1. The API Client, which is essential for all requests and stores crucial data.
2. Models, which define the structure of response bodies.
3. APIs, which are modules offering various functionalities.

Every API requires a specific model to facilitate sending requests. For more information, refer to the documentation at docs.pieces.app/build.
-->

---
transition: fade
---

# API Client Setup

```
import pieces_os_client
import platform

platform_info = platform.platform()

> port = 5323 if 'Linux' in platform_info else 1000

Reviewer: Not needed - the SDK does this automatically.

configuration = pieces_os_client.Configuration(host=f"http://localhost:{port}")
api_client = pieces_os_client.ApiClient(configuration)

```

<!--
Here's how to set up the API Client:
1. We import the necessary modules.

> 2. We determine the correct port based on the operating system.

Reviewer: Done by the SDK

2. We can send messages to the QGPT API stream.
3. The conversation_id can be obtained from the conversations endpoint or left as None to start a new conversation.

This setup allows for interactive, streaming responses from the AI model.
-->

---
transition: fade
layout: center
---

# Let's now build our first application powered by Pieces OS SDK application!

<img class='bottom-image' width='500' src='/build-with-pieces-sdk/img/Robot_Sitting_at_Desk_DM.webp'/>

---
transition: fade
image: '/build-with-pieces-sdk/img/streamlit-light.png'
layout: image-center
imageWidth: '600'
---

# Introduction to Streamlit

Streamlit is an open-source Python framework for data scientists and AI/ML engineers to deliver dynamic data apps with only a few lines of code. You can build and deploy powerful data apps in minutes. 

<!--
Streamlit is an open-source Python framework for data scientists and AI/ML engineers to deliver dynamic data apps with only a few lines of code. You can build and deploy powerful data apps in minutes.
-->

---
transition: fade
layout: center-no-header-footer
disabled: false
---

<SlidevVideo v-click autoplay muted controls width="750">

>   <source src="/build-with-pieces-sdk/video/streamlit-demo.mp4" type="video/mp4">

Reviewer: Link rather than embed - avoids any copyright issues

---
layout: intro
introImage: '/img/pieces-circle.png'
transition: fade
---

# Build AI-powered Apps Using Pieces SDK

Learn how to use the Pieces OS Python SDK

<!--
Welcome to this workshop on learning how to use the pieces_os_client Python SDK.

The pieces_os_client Python SDK is a powerful tool for integrating Pieces OS functionality into your Python applications. This workshop will help you understand and utilize its key features.
-->

---
transition: fade
layout: center

> class: "flex items-center justify-center"

Reviewer: You can remove this - it just sets the top right corner logo as the center, and this is inconsistent with the rest of the slides. Removing it keeps the current content in the center.

<img class='bottom-image' width='400' src='/build-with-pieces-sdk/img/Robot_CTA_Coffee_DM.webp'/>

<!--

Pieces for Developers is an AI-powered tool crafted to assist developers at all levels—from students and hobbyists to seasoned professionals—in streamlining their workflow and enhancing productivity.

Unlike conventional AI programming tools, Pieces is uniquely designed to grasp the full context of your workflow. Whether you're conducting research or writing code in your IDE, Pieces maintains a comprehensive understanding of your activities. This allows you to seamlessly share code snippets, code files, and project folders, and even enables interaction with all the applications running on your computer.

-->

---
transition: fade
---

# What is Pieces OS?

Pieces OS is a critical component of the Pieces productivity suite. It functions as a background service that operates locally on a developer's machine, facilitating communication between various elements of the Pieces ecosystem and managing integrations. This service is essential for running the Pieces applications and any associated plugins.

> <!--->

Reviewer: ```suggestion
<!--
```

- Create a tool that automatically generates documentation based on the codebase and comments. The application can utilize Pieces OS to analyze code and produce structured documentation.

-->
---
transition: fade
---

Infact a few of Pieces extensions/plugins have been developed using the Pieces OS SDK and are completely open source. Here are some including: 

<v-clicks>

- [Pieces Sublime Plugin](https://github.com/pieces-app/plugin_sublime) 
- [Pieces Neovim Plugin](https://github.com/pieces-app/plugin_neo_vim) 
- [Pieces Raycast Extension](https://github.com/raycast/extensions/blob/main/extensions/pieces-raycast)

</v-clicks>

---
transition: fade
---

# Installation of Pieces OS Client SDK for Python

Install the SDK using pip:

```
pip install pieces_os_client
```

<!--
To get started, you'll need to install the SDK using pip. This command will download and install the necessary packages.
-->

---
transition: fade
---

# Core Elements

The `pieces_os_client` SDK consists of:

- **API Client**: Essential for all requests, storing crucial data such as the host.
- **Models**: Define the structure of the response body sent to the Pieces OS API.

> - **APIs**: Modules offering diverse functionalities (e.g., AssetsAPI, QGPTApi).

Reviewer: Better to describe the APIs than name them

```suggestion
- **APIs**: Modules offering diverse functionalities (e.g., APIs for managing assets such as code snippets, interacting with the copilot).
```

```

<!--
Pieces OS Client Python SDK provides you API endpoints to manage your assets. You can create new assets, update and fetch all of your assets using these set of endpoints. 

These endpoints are useful for managing your saved materials inside of Pieces.
-->

---
transition: fade
---

# LLM Management

The Pieces Python SDK provides methods to share list of available LLMs and select the current LLM
```
from pieces_os_client.wrapper import PiecesClient
from pieces_os_client import ClassificationSpecificEnum, FragmentMetadata

# Initialize the PiecesClient
pieces_client = PiecesClient()

# Get all models and print their names
models = pieces_client.get_models()
for model_name, model_id in models.items():
    print(model_name)

# Set the current LLM
pieces_client.model_name = "your_model_name"

# Close the client
pieces_client.close()
```

<!--
The LLM Management APIs are crucial for working with AI models in Pieces OS:
1. We get a snapshot of all available AI models.
2. We can iterate through these models to see their names and IDs.
3. We can select a specific model that we wish to use for Pieces Copilot chats. 

-->

---
transition: fade
---

# Copilot Chats

> The Copilot Chat endpoints provides methods to communicate with Pieces Copilot. You can ask questions to Pieces Copilot using the `stream_question()` method. These methods are powered by the underlying QGPT endpoints. 

Reviewer: QGPT is an internal term, so maybe pick something clearer - or explain this is the name of the underlying API

1. The stream_question() requires a question as a parameter and will stream the response.
2. The chats() method returns a list of all chats.
3. You can fetch messages from a specific chat add further messages to a given chat

This setup allows for interactive, streaming responses from the AI model.
-->

---
transition: fade
layout: center
---

# Let's now build our first application powered by Pieces OS SDK application!

<img class='bottom-image' width='300' src='/build-with-pieces-sdk/img/Robot_Sitting_at_Desk_DM.webp'/>

---
transition: fade
image: '/build-with-pieces-sdk/img/streamlit-light.png'
layout: image-center
imageWidth: '600'
---

# Introduction to Streamlit

Streamlit is an open-source Python framework for data scientists and AI/ML engineers to deliver dynamic data apps with only a few lines of code. You can build and deploy powerful data apps in minutes. 

<!--
Streamlit is an open-source Python framework for data scientists and AI/ML engineers to deliver dynamic data apps with only a few lines of code. You can build and deploy powerful data apps in minutes.
-->

---
transition: fade
layout: center

> class: "flex items-center justify-center"

Reviewer: Not needed

---

<SlidevVideo v-click autoplay muted controls width="750" class="m-auto">
  <source src="https://s3-us-west-2.amazonaws.com/assets.streamlit.io/videos/hero-video.mp4" type="video/mp4">

> </SlidevVideo>0

Reviewer: ```suggestion
</SlidevVideo>
```
